src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -69,15 +76,33 @@ internal sealed class VisualStudioProject
         /// </summary>
         private readonly FileChangeWatcher.IContext _fileReferenceChangeContext;
 
+        /// <summary>
+        /// track whether we have been subscribed to <see cref="IDynamicFileInfoProvider.Updated"/> event
+        /// </summary>
+        private readonly HashSet<IDynamicFileInfoProvider> _eventSubscriptionTracker = new HashSet<IDynamicFileInfoProvider>();
+
+        /// <summary>
+        /// map original dynamic file path to <see cref="DynamicFileInfo.FilePath"/>
+        /// </summary>
+        private readonly Dictionary<string, string> _dynamicFilePathMaps = new Dictionary<string, string>();
+
         private readonly BatchingDocumentCollection _sourceFiles;
         private readonly BatchingDocumentCollection _additionalFiles;
 
         public ProjectId Id { get; }
         public string Language { get; }
 
-        internal VisualStudioProject(VisualStudioWorkspaceImpl workspace, HostDiagnosticUpdateSource hostDiagnosticUpdateSource, ProjectId id, string projectUniqueName, string language, string directoryNameOpt)
+        internal VisualStudioProject(
+            VisualStudioWorkspaceImpl workspace,
+            ImmutableArray<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> dynamicFileInfoProviders,
+            HostDiagnosticUpdateSource hostDiagnosticUpdateSource,
+            ProjectId id,
+            string projectUniqueName,
+            string language,
+            string directoryNameOpt)
```

Comment:
is changing this signature going to be a problem for TS/JS?

Reply:
@jasonmalinowski this is new type jason just added. so probably not used by anyone else. at least yet.